teps/0011-redirecting-step-output-streams.md

```diff
@@ -0,0 +1,218 @@
+
+* Allow tool developers to create CI pipeline tooling that can use Task results on *any* image specified by the end users without the above limitations. As Task authors, one can choose which images to use to work around the limitations. But as a tool developer, one cannot or should not control what images their end users can use, and cannot assume any details about the images.
+
+### Goals
+
+1. Allow the stdout/stderr of a step to be consumed by another step.
+1. Enable a user to configure the path where the output streams are written.
+
+### Non-Goals
+
+1. Parse stdout/stderr of a step into a structured format (e.g., JSON) and extract information from certain fields.
+
+## Requirements
+
+* Add new fields for a step to specify paths to redirect stdout/stderr to.
+* Users should be able to observe the output streams through Pod logs even if stdout/stderr redirections are specified. In other words, output streams should be duplicated instead of simply redirected.
+* Clearly documents the restrictions when stdout/stderr redirections are set to Task result paths, and encourage Task authors to set the paths to workspace paths, especially if they want to exchange large amount of data between Tasks.
+* Provide examples to use step redirection for the use cases mentioned in [Motivation](#motivation).
+
+## Proposal
+
+### Option 1: Allowing Users to Specify Redirection Paths
+
+The following new fields will be added to the `Step` struct:
+
+```go
+// StepOutputConfig stores configuration for a step output stream.
+type StepOutputConfig {
+  // Path to duplicate stdout stream to on container's local filesystem.
+  // +optional
+  Path string `json:"path,omitempty"`
+}
+
+type Step struct {
+    ...
+    // Stores configuration for the stdout stream of the step.
+    StdoutConfig StepOutputConfig `json:"stdoutConfig"`
+    // Stores configuration for the stderr stream of the step.
+    StderrConfig StepOutputConfig `json:"stderrConfig"`
+}
+```
+
+Once `StdoutConfig.Path` or `StderrConfig.Path` is specified, the corresponding output stream will be duplicated to both the given file and the standard output stream of the container, so users can still view the output through the Pod log API. If both `StdoutConfig.Path` and `StderrConfig.Path` are set to the same value, the two streams will be written to two file descriptors referring to the same file description, so outputs from both stream will be interleaved in the same file, but there will be no ordering guarantee on the data.  If multiple steps' `StdoutConfig.Path` fields are set to the same value, the file content will be overwritten by the last outputting step.
```

Comment:
> the two streams will be written to two file descriptors referring to the same file description,

Not quire sure about the wording, something like: 'written to two file descriptors which are both open for write on the same file'?

Reply:
There's a subtle difference. When you open a file through `open` syscall, you create a new file description entry in the operating system's global file description table, and a file descriptor that points to the entry; when you duplicate a file descriptor through `dup2` syscall, you create a new file descriptor that points to the same entry in the file description table. The operating system (usually) ensures consistency for all read/write to a a file description entry; even if the read/write are done through different file descriptors; but each descriptor maintains its own flags and cursor.

This is probably just an implementation detail so I'll just remove this sentence.

Reply:
ah kk no probs! either removing the sentence or just adding in your explanation above sgtm :) Thanks for explaining